src/app/CommandSender.cpp

```diff
@@ -409,6 +429,27 @@ CHIP_ERROR CommandSender::ProcessInvokeResponseIB(InvokeResponseIB::Parser & aIn
         }
         ReturnErrorOnFailure(err);
 
+        if (commandRef.HasValue() && mpPendingResponseTracker != nullptr)
+        {
+            err = mpPendingResponseTracker->ResponseReceived(commandRef.Value());
+            if (err != CHIP_NO_ERROR)
+            {
+                // This can happen for two reasons:
+                // 1. The current InvokeResponse is a duplicate (based on its commandRef).
+                // 2. The current InvokeResponse is for a request we never sent (based on its commandRef).
+                // Used when logging errors related to server violating spec.
+                [[maybe_unused]] ScopedNodeId remoteScopedNode;
+                if (mExchangeCtx.Get()->HasSessionHandle())
+                {
+                    remoteScopedNode = mExchangeCtx.Get()->GetSessionHandle()->GetPeer();
+                }
+                ChipLogError(DataManagement,
+                             "Received Unexpected Response from remote node " ChipLogFormatScopedNodeId ", commandRef=%u",
+                             ChipLogValueScopedNodeId(remoteScopedNode), commandRef.Value());
+            }
+            ReturnErrorOnFailure(err);
```

Comment:
nit: this could be moved up as a return in the above if (since it checks for an error)